src/engine/GroupBy.h

```diff
@@ -190,11 +192,97 @@ class GroupBy : public Operation {
+          val != nullptr && val->isUndefined()) {
+        currentValue_ = value;
+        return;
+      }
+      auto impl = [&ctx]<typename X, typename Y>(const X& x, const Y& y) {
+        return toBoolNotUndef(
+            sparqlExpression::compareIdsOrStrings<
+                Comp, valueIdComparators::ComparisonForIncompatibleTypes::
+                          CompareByType>(x, y, ctx));
+      };
+      if (std::visit(impl, value, currentValue_)) currentValue_ = value;
+    }
+    [[nodiscard]] ValueId calculateResult(LocalVocab* localVocab) const {
+      if (auto val = std::get_if<ValueId>(&currentValue_))
+        return *val;
+      else if (auto str = std::get_if<std::string>(&currentValue_)) {
+        auto localVocabIndex = localVocab->getIndexAndAddIfNotContained(*str);
+        return ValueId::makeFromLocalVocabIndex(localVocabIndex);
+      }
+      return ValueId::makeUndefined();
+    }
+  };
+
+  using MinAggregationData =
+      ExtremumAggregationData<valueIdComparators::Comparison::LT>;
+  using MaxAggregationData =
+      ExtremumAggregationData<valueIdComparators::Comparison::GT>;
+
+  // Data to perform the SUM aggregation using the HashMap optimization.
+  struct SumAggregationData {
+    using ValueGetter = sparqlExpression::detail::NumericValueGetter;
+    bool error_ = false;
+    double sum_ = 0;
+    void increment(auto&& value,
+                   const sparqlExpression::EvaluationContext* ctx) {
+      auto val = ValueGetter{}(AD_FWD(value), ctx);
+
+      if (const int64_t* intval = std::get_if<int64_t>(&val))
+        sum_ += static_cast<double>(*intval);
+      else if (const double* dval = std::get_if<double>(&val))
+        sum_ += *dval;
+      else
+        error_ = true;
+    };
+    [[nodiscard]] ValueId calculateResult(const LocalVocab* localVocab) const {
+      (void)localVocab;
+      if (error_)
+        return ValueId::makeUndefined();
+      else
+        return ValueId::makeFromDouble(sum_);
+    }
+  };
+
+  // Data to perform GROUP_CONCAT aggregation using the HashMap optimization.
+  struct GroupConcatAggregationData {
+    using ValueGetter = sparqlExpression::detail::StringValueGetter;
+    std::string currentValue_;
+    std::string seperator_;
+    void increment(auto&& value,
+                   const sparqlExpression::EvaluationContext* ctx) {
+      auto val = ValueGetter{}(AD_FWD(value), ctx);
+      if (val.has_value()) {
+        if (!currentValue_.empty()) currentValue_.append(seperator_);
+        currentValue_.append(val.value());
+      }
+    }
+    [[nodiscard]] ValueId calculateResult(LocalVocab* localVocab) const {
+      auto localVocabIndex =
+          localVocab->getIndexAndAddIfNotContained(currentValue_);
+      return ValueId::makeFromLocalVocabIndex(localVocabIndex);
+    }
+
+    explicit GroupConcatAggregationData(std::string seperator)
+        : seperator_{std::move(seperator)} {
+      currentValue_.reserve(20000);
+    }
+  };
```

Comment:
I think with this PR you can start moving these aggregationData classes to a separate header to net let the files become too big.